File: 3020-find-the-maximum-number-of-elements-in-subset/3020-find-the-maximum-number-of-elements-in-subset.py
class Solution(object):
    def maximumLength(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        freq = Counter(nums)
        ans = 0;

        for key, val in freq.items():
            curNum = key

            if curNum == 1:
                oneLength = val -1 if val%2 == 0 else val
                ans = max(ans, oneLength)
                continue
            
            curLength = 0

            while curNum in freq and freq[curNum] >=2:
                curLength = curLength + 2
                curNum = curNum * curNum
            
            if curNum in freq and freq[curNum] == 1:
                curLength += 1
            else:
                curLength -= 1
            
            ans = max(ans, curLength)

        return ans

File: 3020-find-the-maximum-number-of-elements-in-subset/3020-find-the-maximum-number-of-elements-in-subset.cpp
class Solution {
public:
    int maximumLength(vector<int>& nums) {
        unordered_map<long long, long long> freq;
        long long ans = 0;

        for (auto& n : nums)
            freq[n]++;

        for (auto& x : freq) {
            long long curNum = x.first;

            // special case
            if (curNum == 1) {
                ans = max(ans, (freq[1] % 2 == 0 ? freq[1] - 1 : freq[1]));
                continue;
            }

            long long curLength = 0;

            while (freq.count(curNum) && freq[curNum] >= 2) {
                curLength += 2;
                curNum *= curNum;
            }

            if (freq.count(curNum) && freq[curNum] == 1)
                curLength++;
            else
                curLength--;

            ans = max(ans, curLength);
        }

        return ans;
    }
};